schema/schema.go

```diff
@@ -19,49 +20,92 @@ func setup(tx *sql.Tx) error {
+	if err = applySetup(c); err != nil {
+		return err
 	}
 
 	var migrations = ReadMigrations(root)
 	sort.Stable(migrations) // sort in ascending order of version number
 
-	var currentVersion = 0
-	const fetchCurrentVersion = "SELECT version FROM sqlq_migrations ORDER BY applied_on DESC, version DESC LIMIT 1"
+	currentVersion, err := getCurrentMigrationVersion(c)
+	if err != nil {
+		return err
+	}
+
+	for _, mg := range migrations {
 
+		if mg.Version() <= currentVersion {
+			continue // already applied. skip this migration
+		}
+
+		if err = applyMigrations(c, mg); err != nil {
+			return errors.Wrapf(err, "failed to apply migration: name=%s\tversion=%d", mg.Name(), mg.Version())
+		}
+
+	}
+
+	return nil
+}
+
+// getCurrentMigrationVersion gets the current version of our migrations
+func getCurrentMigrationVersion(c *sql.DB) (int, error) {
+	const fetchCurrentVersion = "SELECT version FROM sqlq_migrations ORDER BY applied_on DESC, version DESC LIMIT 1"
+	var currentVersion int
+	var err error
 	var rows *sql.Rows
-	if rows, err = tx.Query(fetchCurrentVersion); err != nil && err != sql.ErrNoRows {
-		return errors.Wrap(err, "failed to fetch latest migration version")
+
+	if rows, err = c.Query(fetchCurrentVersion); err != nil && err != sql.ErrNoRows {
+		return 0, errors.Wrap(err, "failed to fetch latest migration version")
 	}
 
 	if rows.Next() {
 		if err = rows.Scan(&currentVersion); err != nil && err != sql.ErrNoRows {
-			return errors.Wrap(err, "failed to fetch latest migration version")
+			return 0, errors.Wrap(err, "failed to fetch latest migration version")
 		}
 	}
 
-	for _, mg := range migrations {
-		if mg.Version() <= currentVersion {
-			continue // already applied. skip this migration
-		}
+	if err = rows.Err(); err != nil {
+		return 0, errors.Wrap(err, "failed to iterate rows")
+	}
 
-		if err = mg.Apply(tx); err != nil {
-			return errors.Wrapf(err, "failed to apply migration(%s)", mg.Name())
-		}
+	return currentVersion, errors.Wrapf(err, "failed getting current migration version")
+}
 
-		const recordMigration = "INSERT INTO sqlq_migrations(name, version) VALUES ($1, $2)"
-		if _, err = tx.Exec(recordMigration, mg.Name(), mg.Version()); err != nil {
-			return errors.Wrapf(err, "failed to record migration: name=%s\tversion=%d", mg.Name(), mg.Version())
-		}
+// applySetup creates if needed a migrations table
+func applySetup(c *sql.DB) error {
+	var tx *sql.Tx
+	var err error
+	if tx, err = c.Begin(); err != nil {
+		return errors.Wrapf(err, "failed to start new transaction")
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	if err = setup(tx); err != nil {
+		return errors.Wrapf(err, "failed to perform setup")
 	}
+	return errors.Wrapf(tx.Commit(), "failed to commit setup")
+}
 
-	return errors.Wrapf(tx.Commit(), "failed to commit migration")
+// applyMigrations is responsible  for apply a migration individually
+func applyMigrations(c *sql.DB, mg *EmbeddedMigration) error {
```

Comment:
```suggestion
// applyMigration applies a single embedded migration to the database
func applyMigration(c *sql.DB, mg *EmbeddedMigration) error {
```

Just a naming change to make it clear that it only runs one migration, not all of them